test/session-pool.js

```diff
@@ -17,734 +17,489 @@
+    it('should set isOpen to true', function() {
+      sessionPool.open();
+      assert.strictEqual(sessionPool.isOpen, true);
     });
 
-    it('should clear the read and write pools', function() {
-      sessionPool.writePool = {
-        drain: function() {
-          sessionPool.writePool.drain.called = true;
-          return Promise.resolve();
-        },
-        clear: function() {
-          sessionPool.writePool.clear.called = true;
-          return Promise.resolve();
-        },
-      };
+    it('should start listening for events', function(done) {
+      sessionPool.listenForEvents_ = done;
+      sessionPool.open();
+    });
 
-      return sessionPool.clear().then(function() {
-        assert(sessionPool.pool.drain.called);
-        assert(sessionPool.pool.clear.called);
-        assert(sessionPool.writePool.drain.called);
-        assert(sessionPool.writePool.clear.called);
-      });
+    it('should emit an open event', function(done) {
+      sessionPool.once('open', done);
+      sessionPool.open();
     });
-  });
 
-  describe('getSession', function() {
-    it('should get next available if pool is not free', function(done) {
-      sessionPool.pool = {
-        free: false,
-      };
+    it('should create the min. of readonly sessions', function() {
+      var min = (sessionPool.options.min = 10);
+      var createCallCount = 0;
 
-      sessionPool.getNextAvailableSession_ = function(callback) {
-        callback(); // done()
+      sessionPool.createSessionInBackground_ = function(type) {
+        assert.strictEqual(type, 'readonly');
+        createCallCount += 1;
       };
 
-      sessionPool.getSession(done);
+      sessionPool.open();
+
+      assert.strictEqual(createCallCount, min);
     });
 
-    it('should acquire from the pool', function(done) {
-      var session = {};
+    it('should factor in the number of desired write sessions', function() {
+      var min = (sessionPool.options.min = 10);
+      var writes = (sessionPool.options.writes = 5);
 
-      function callback(err, session_) {
-        assert.ifError(err);
-        assert.strictEqual(session_, session);
-        done();
-      }
+      var callCounts = {
+        readonly: 0,
+        readwrite: 0,
+      };
 
-      sessionPool.pool = {
-        free: true,
-        acquire: function() {
-          return Promise.resolve(session);
-        },
+      sessionPool.createSessionInBackground_ = function(type) {
+        callCounts[type] += 1;
       };
 
-      sessionPool.getSession(callback);
-    });
+      sessionPool.open();
 
-    it('should return error when it cannot acquire the pool', function(done) {
-      var error = new Error('Error.');
+      assert.strictEqual(callCounts.readonly, min - writes);
+      assert.strictEqual(callCounts.readwrite, writes);
+    });
 
-      function callback(err) {
-        assert.strictEqual(err, error);
-        done();
-      }
+    it('should settle once all sessions are created', function() {
```

Comment:
Does this mean that session pool is only usable after all of the min sessions have been created. If yes then that is not idle. It should be usable immediately and any getSession would block till there is any session available.

Reply:
Nope! We don't want for the pool when using a Database object, they can use it immediately and unless configured to fail, it would wait for a session to free up.